zkevm-circuits/src/tx_circuit/sign_verify.rs

```diff
@@ -0,0 +1,978 @@
+use gadgets::is_zero::{IsZeroChip, IsZeroConfig, IsZeroInstruction};
+use group::{ff::Field, prime::PrimeCurveAffine, Curve};
+use halo2_proofs::{
+    arithmetic::{BaseExt, Coordinates, CurveAffine, FieldExt},
+    circuit::{AssignedCell, Layouter, Region},
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Selector},
+    poly::Rotation,
+};
+use integer::{
+    AssignedInteger, IntegerChip, IntegerConfig, IntegerInstructions, WrongExt,
+    NUMBER_OF_LOOKUP_LIMBS,
+};
+use itertools::Itertools;
+use keccak256::plain::Keccak;
+use lazy_static::lazy_static;
+use log::error;
+use maingate::{
+    Assigned, AssignedValue, MainGate, MainGateConfig, MainGateInstructions, RangeChip,
+    RangeConfig, RangeInstructions, RegionCtx, UnassignedValue,
+};
+use secp256k1::Secp256k1Affine;
+use std::{convert::TryInto, io::Cursor, marker::PhantomData};
+
+/// Power of randomness vector size required for the SignVerifyChip
+pub const POW_RAND_SIZE: usize = 63;
+
+/// Number of rows required for a verification of the SignVerifyChip in the
+/// "signature address verify" region.
+pub const VERIF_HEIGHT: usize = 1;
+
+/// Auxiliary Gadget to verify a that a message hash is signed by the public
+/// key corresponding to an Ethereum Address.
+#[derive(Default, Debug)]
+pub(crate) struct SignVerifyChip<F: FieldExt, const MAX_VERIF: usize> {
+    pub aux_generator: Secp256k1Affine,
+    pub window_size: usize,
+    pub _marker: PhantomData<F>,
+}
+
+const KECCAK_IS_ENABLED: usize = 0;
+const KECCAK_INPUT_RLC: usize = 1;
+const KECCAK_INPUT_LEN: usize = 2;
+const KECCAK_OUTPUT_RLC: usize = 3;
+
+const NUMBER_OF_LIMBS: usize = 4;
+const BIT_LEN_LIMB: usize = 72;
+
+/// Return an expression that builds an integer element in the field from the
+/// `bytes` in little endian.
+fn int_from_bytes_le<'a, F: FieldExt>(
+    bytes: impl IntoIterator<Item = &'a Expression<F>>,
+) -> Expression<F> {
+    // sum_{i = 0}^{N} bytes[i] * 256^i
+    let mut res = 0u8.expr();
+    for (i, byte) in bytes.into_iter().enumerate() {
+        res = res + byte.clone() * Expression::Constant(F::from(256).pow(&[i as u64, 0, 0, 0]))
+    }
+    res
+}
+
+/// Enable copy constraint between `src` integer limbs and `dst` limbs.  Then
+/// assign the `dst` limbs values from `src`.
+fn copy_integer<F: FieldExt, W: WrongExt>(
+    region: &mut Region<'_, F>,
+    name: &str,
+    src: AssignedInteger<W, F, NUMBER_OF_LIMBS, BIT_LEN_LIMB>,
+    dst: &[Column<Advice>; 4],
+    offset: usize,
+) -> Result<(), Error> {
+    for (i, limb) in src.limbs().iter().enumerate() {
+        let assigned_cell = region.assign_advice(
+            || format!("{} limb {}", name, i),
+            dst[i],
+            offset,
+            || limb.value().ok_or(Error::Synthesis),
+        )?;
+        region.constrain_equal(assigned_cell.cell(), limb.cell())?;
+    }
+    Ok(())
+}
+
+/// Enable copy constraints between `src` integer bytes and `dst` integer bytes.
```

Comment:
Copy constraints are enabled AFAIK with `.enable_equality`, so to avoid confusions maybe better just to say thay copies constraints.

Reply:
Thanks! I'll reword it to:
```rust
/// Constraint equality (using copy constraints) between `src` integer bytes and
/// `dst` integer bytes. Then assign the `dst` values from `src`.
```